chore(dynamo-run): Refactor to library

grahamking opened the pull request:
Move much of what was in the `dynamo-run` crate into `dynamo-llm` so that everyone can use it.

Example usage:

1. Create a `LocalModel`:

```
    let local_model = LocalModelBuilder::default()
	.model_path("Qwen/Qwen3-0.6B")
	.http_port(8080)
	.build().await?;
```

2. Make an engine:

```
    let engine_config = EngineConfig::StaticFull {
	engine: dynamo_engine_mistralrs::make_engine(&local_model).await?,
	model: Box::new(local_model),
    };
```

3. Connect it to an input and run it

```
    dynamo_llm::entrypoint::input::run_input(Input::Http, runtime, engine_config).await?;
```

For https://github.com/ai-dynamo/dynamo/issues/1647


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced a flexible builder pattern for local model configuration, allowing advanced customization and easier initialization.
  * Added new input modes and unified input handling, supporting interactive chat, HTTP server, batch file, and distributed endpoint modes.
  * Centralized engine configuration and routing, enabling more extensible and maintainable engine management.

* **Refactor**
  * Simplified and modularized the codebase by moving input and engine logic into dedicated modules.
  * Replaced direct model construction with an asynchronous builder for improved clarity and extensibility.
  * Streamlined configuration and validation for flags and router settings.

* **Bug Fixes**
  * Added validation to prevent incompatible input and output combinations in endpoint and dynamic modes.

* **Chores**
  * Updated and reorganized dependencies to reflect new input and configuration features.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

grahamking committed:
dynamo-run: Refactor to library

Move much of what was in the `dynamo-run` crate into `dynamo-llm` so
that everyone can use it.

Example usage:

1. Create a `LocalModel`:

```
    let local_model = LocalModelBuilder::default()
	.model_path("Qwen/Qwen3-0.6B")
	.http_port(8080)
	.build().await?;
```

2. Make an engine:

```
    let engine_config = EngineConfig::StaticFull {
	engine: dynamo_engine_mistralrs::make_engine(&local_model).await?,
	model: Box::new(local_model),
    };
```

3. Connect it to an input and run it

```
    dynamo_llm::entrypoint::input::run_input(Input::Http, runtime, engine_config).await?;
```

For https://github.com/ai-dynamo/dynamo/issues/1647

grahamking committed:
fix: Code Rabbit is wonderful

And fix bindings, update lock files.

paulhendricks reviewed (approved):
Great work! Left a few little comments.

grahamking commented on lib/llm/src/entrypoint.rs:
```diff
@@ -0,0 +1,60 @@
+    },
+}
+
+impl EngineConfig {
+    fn local_model(&self) -> &LocalModel {
+        use EngineConfig::*;
+        match self {
+            Dynamic(lm) => lm,
+            StaticFull { model, .. } => model,
+            StaticCore { model, .. } => model,
+        }
+    }
+}
```
Nope, just personal preference.

grahamking committed:
Change type block_size and context_length from usize to u32

They were usize because that's easiest in Rust, but naturally they are
u32, so put in the extra work to do it right.

The nice part is at the boundary (mistralrs, gguf, etc) they were already u32, suggesting
this is indeed the correct type.

grahamking commented on lib/llm/src/local_model.rs:
```diff
@@ -25,58 +30,85 @@ const HF_SCHEME: &str = "hf://";
+    router_config: Option<RouterConfig>,
+    kv_cache_block_size: usize,
+    http_port: u16,
 }
```
Done. Phew, turns out `block_size` is *everywhere*.

grahamking committed:
Fix types in Python bindings for kv_cache_block_size and context_length
